kernel: Make If statement have essentially Boolean type

ceolin opened the pull request:
Make if statement using pointers explicitly check whether the value is
NULL or not.

The C standard does not say that the null pointer is the same as the
pointer to memory address 0 and because of this is a good practice
always compare with the macro NULL.

part of #9889

Commit:
kernel: Make If statement have essentially Boolean type

Make if statement using pointers explicitly check whether the value is
NULL or not.

The C standard does not say that the null pointer is the same as the
pointer to memory address 0 and because of this is a good practice
always compare with the macro NULL.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andyross commented:
OK, I'm going to push back a bit on this one, because IMHO this particular convention is a good thing and not a bad thing.  What exactly does MISRA say here?  Because I'm all but certain ISO C does, in fact, guarantee that NULL, however defined (which is to say, defined as zero, because no one has ever used a non-zero NULL on a shipping system), will evaluate to a boolean false.  C++ did this by defining "0" as a pointer constant for a while, though they've since moved to that "nullptr" thing as official.

Honestly I think these added bytes hurt readability and don't help, especially when combined with loop iteration and while() tests with side effects.

ceolin commented:
c standard is not clear about NULL definition, in 6.3.2.3 it says:
```
An integer constant expression with the value 0, or such an expression
cast to type void *, is called a null pointer constant. 55) If a null
pointer constant is converted to a pointer type, the resulting pointer,
called a null pointer, is guaranteed to compare unequal to a pointer to
any object or function.
```

but in 7.17:
```
The macros are NULL
which expands to an implementation-defined null pointer constant;
```

In other hand MISRA-C just says:
```
The controlling expression of a for statement is optional. The rule
does not re quire the expression to be present but does require it to have essentially Boolean type if it is present.
```

IMHO it increase code readability too, I know that we have to type more, but when you see something like `if (var == 0) { ... }` will automatically know that `var` is an integer and when is `if (var == NULL) { ...}` you know that `var` is a pointer.

ceolin commented:
another thing, this guideline is not only to pointers, but all `if` statements should evaluate a Boolean expression so, we'll need to change the others types types too, integers should be compared against constants .... The only exception is:
```
bool t = true;
...
if (t) {
...
}
```

andyross commented:
> will automatically know that var is an integer 

That's the logic that gave us hungarian notation. :)

In fact variable typing isn't really a big limit to readability.  Computers check types for us automatically, and we all learn very quickly to read code with the assumption that typing is correct.  What we really want to know in code like that is "the variable is VALID", which is IMHO better conveyed by simply using its name in a boolean context.

But whatever.  I'm not going to put up much of a fuss.  This is just a spot where I genuinely feel we're moving backwards.

ceolin commented:
There is this ambiguity in C (at least 90 and 99) standard, I know that every sane implementation define NULL to `((void *)0)`, but is always better strict follow the standard when is something that has no harmful effect. It definitely helps portability.

Just my two penny worth :)

pfalcon commented on kernel/poll.c:
```diff
@@ -80,8 +80,9 @@ static inline void add_event(sys_dlist_t *events, struct k_poll_event *event,
 	pending = (struct k_poll_event *)sys_dlist_peek_tail(events);
-	if (!pending || _is_t1_higher_prio_than_t2(pending->poller->thread,
-						   poller->thread)) {
+	if ((pending == NULL) ||
```
Superfluous parens.

ceolin replied:
this case for me increase readability :)

I'm normalizing this for:

```
if ((x == NULL) && (y != MAX_INT)) {...}
if (x != NULL) {...}
if ((x == NULL) && _some_function_returning_boolean_constant()) {...}
```

Add parentheses around the function call is weird and decide to not add parentheses around the first expression will decrease readability and ...
most cases already have parenthesis around

pfalcon commented on kernel/queue.c:
```diff
@@ -303,7 +303,7 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 		irq_unlock(key);
 
-		if (!val && timeout != K_FOREVER) {
+		if ((val == NULL) && timeout != K_FOREVER) {
```
This isn't just superfluous, but inconsistent parens. We can't really use parens on left side of && and not use on right. (It should be consistent - and none required left or right).

ceolin replied:
ok, I can do that. In this case I have created a script to do these changes.
I can manually change this occurrence or change the script to enforce parentheses in the other expressions.

For now I'll just change this expression as you asked.

ceolin replied:
@pfalcon got confuse now. Most zephyr's code is doing

```
if (x && (y != K_SOMETHING)) { ... }
```

This case is not the most common, it's easier changing adding the parentheses int the second expression. One thing that MISRA-C tries to do is decrease the chance of a developer introduce errors, in this case the parentheses is something valid IMHO because there are less chance of a programmer introduce an error caused by precedence of operators.


That's my opinion, I'm totally open to remove the parentheses in all cases, I just prefer send the fix prior bring a lot of discussion that normally takes infinity time. We need to move forward in this MISRA task :)

pfalcon replied:
> Most zephyr's code is doing
> if (x && (y != K_SOMETHING)) { ... }

I did `grep " && " *` in kernel/ , and word "most" is not which I'd use to describe the situation. Indeed, both extra-parens and no-parens styles are used throughout. My concern here specifically was inconsistent use of parens one side vs another side of operator, but as we can see, the problem is deeper than that, definitely not to address this time. Let me dismiss my -1 so just current task at hand can proceed. Thanks for paying attention to matters which arise around addressing the MISRA matters.

pfalcon reviewed (dismissed):
This is the change (almost the only one brought by MISRA) which I can whole-heartedly agree with.

But please try not just formally apply changes for MISRA, but actually improve readability of the code. And changes like:

~~~
+if ((val == NULL) && timeout != K_FOREVER)
~~~

with superfluous parentheses introduced on the left side of && while non-parens subexpression stays on the right side - is hardly consistent/good looking. In this and other cases extra parens are not needed.

Commit:
kernel: Using boolean constants instead of 0 or 1

MISRA C requires that every controlling expression of and if or while
statement have a boolean type.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

pfalcon commented:
> OK, I'm going to push back a bit on this one

My 2 cents: I wish there was push back on (void)memset ;-). As an example, on my recent memory (2-3 months), there were 2 fixes in Zephyr, when `r` was used instead of `!r` and/or vice-versa.

I have some experience recent years working with a codebase which enforces this rule (which can be seen as MISRA'esque, or otherwise as Java'esque), and have to admit it makes the code clearer/chance of mistakes lower. (As a full disclosure, I hate `5 == foo` style, that's too much IMHO. But `p != NULL` is ok.)

Commit:
kernel: Compare pointers with NULL in while statements

Make while statement using pointers explicitly check whether
the value is NULL or not.

The C standard does not say that the null pointer is the same
as the pointer to memory address 0 and because of this is a good
practice always compare with the macro NULL.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@pfalcon hungarian notation hurts :) 

To my mind a lot of these MISRA-C rules make sense in general, they increase portability and make the code less error prone. The problem is that we have  a lot of bad API that cause these kind of patches :/

ceolin commented:
@pfalcon I update this patch adding the missing parentheses in the right expression. For the sake of readability if you guys prefer I can change it removing parentheses in both expressions (will increase the patch I guess :) )

I've added other similar patch  for while() statement, and another one using boolean constants instead of 0 or 1. The latter I can remove from this pr if we need to discuss it separately.

Thanks for reviewing it :)

pfalcon commented:
> I've added other similar patch for while() statement, and another one using boolean constants instead of 0 or 1. The latter I can remove from this pr if we need to discuss it separately.

I'm all for using boolean expressions in boolean contexts and for boolean func params. +1 from my side.

ceolin commented:
@nashif can we merge it ?
